### PR TITLE
[C++ API] support loading C++ dynamic libraries from code search path

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

- Rename the config name from "dynamic_library_path" to "code_search_path". You can set this like "/path1/xxx.so:/path2".
- Remove the 'lib_name' field in task spec. Use "code_search_path" in JobConfig to search shared libraries instead.
- If no "code_search_path" provided, use the program location path by default.
- Load libraries during worker starting, instead of task execution.

## Related issue number
#16735

### CODE REVIEW

File: cpp/src/ray/util/function_helper.cc
Comment (qicosmos) [inline cpp/src/ray/util/function_helper.cc:66]: LoadAllRemoteFunctions. -> GetAllRemoteFunctionNames

File: cpp/src/ray/util/function_helper.cc
Comment (qicosmos) [reply cpp/src/ray/util/function_helper.cc:133]: I think we should make the code more clear and simple, remove the argument : const void *, and we can divide the functions into two functions, one is for normal task, the other is for actor, then the code  will be clear and simple.

File: cpp/src/ray/runtime/task/task_executor.cc
Comment (qicosmos) [inline cpp/src/ray/runtime/task/task_executor.cc:90]: As commend at the function GetExecutableFunctions, here we can explicit call the get GetExecutableFunction and GetExecutableMemberFunction would be very clear and easy to read.

File: cpp/src/ray/util/function_helper.cc
Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.cc:20]: OK.

File: cpp/src/ray/util/function_helper.cc
Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.cc:66]: I think `LoadAllRemoteFunctions` is more reasonable because this function will load the function maps from the library. And the return value of function names is secondary.

File: cpp/src/ray/util/function_helper.h
Comment (qicosmos) [inline cpp/src/ray/util/function_helper.h:37]: Why can't find the function name? i think we register all functions, if we can't find the function, it should be an invalid request, and we need to handle the exception, don't return nullptr.

File: cpp/src/ray/util/function_helper.h
Comment (qicosmos) [inline cpp/src/ray/util/function_helper.h:42]: If we can't find the path, we should do error handling, don't return nullptr.

File: cpp/src/ray/util/function_helper.h
Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.h:37]: If you has multiple dynamic library, some map finding will return `nullptr`.

Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.h:37]: And I have throw an exception here https://github.com/ray-project/ray/pull/16828/files#diff-d7517919ec685a9ab470810f835ec5912fae4699db5d5494223a5278b4e9caf4R160.

File: cpp/BUILD.bazel
Comment (kfstorm) [inline cpp/BUILD.bazel:153]: Do we still need this argument for tests?

File: cpp/include/ray/api/ray_config.h
Comment (kfstorm) [inline cpp/include/ray/api/ray_config.h:22]: We should create a `job_config` field and put `code_search_path` into it. We may add more job config options in the future.

File: cpp/src/ray/config_internal.cc
Comment (kfstorm) [inline cpp/src/ray/config_internal.cc:13]: Could you copy the comment from `gcs.proto`?

File: cpp/src/ray/config_internal.cc
Comment (kfstorm) [inline cpp/src/ray/config_internal.cc:64]: You missed the last segment if there are multiple paths. Have you tested multiple paths?

```suggestion
      while (start < FLAGS_ray_code_search_path.size()) {
        auto end = FLAGS_ray_code_search_path.find(separator, start);
        if (end == std::string::npos) {
          end = FLAGS_ray_code_search_path.size();
        }
        if (start < end) { // In case there are unnecessary separators.
          code_search_path.emplace_back(FLAGS_ray_code_search_path.substr(start, end - start);
        }
        start = end + 1;
      }
```

File: cpp/src/ray/config_internal.cc
Comment (kfstorm) [inline cpp/src/ray/config_internal.cc:100]: Please make it not a warning because it's totally valid use.

File: cpp/src/ray/config_internal.h
Comment (kfstorm) [inline cpp/src/ray/config_internal.h:27]: Make it a sub field. RayConfig -> job_config -> code_search_path.

File: src/ray/raylet/worker_pool.cc
Comment (kfstorm) [inline src/ray/raylet/worker_pool.cc:221]: nit.

```suggestion
        std::string code_search_path_str;        
        if (language == Language::JAVA) {
          code_search_path_str = "-Dray.job.code-search-path=" + code_search_path_str;
        } else if (language == Language::CPP) {
          code_search_path_str = "--ray-code-search-path=" + code_search_path_str;
        } else {
          RAY_LOG(FATAL) << "Unknown language " << Language_Name(language);
        }
        options.push_back(code_search_path_str);
```

**Comment by kfstorm** [comment]
I didn't review library loading, function finding and execution part. I think @qicosmos already did a thorough review.

File: cpp/BUILD.bazel
Comment (SongGuyang) [reply cpp/BUILD.bazel:153]: Yes, we need to make sure this argument works. And the argument of example test has been removed. https://github.com/ray-project/ray/pull/16828/files#diff-1706c14d71810b14bd8b98c70f7651d2ea3f7f50225d2e2d352759251ce24ab5L7

File: cpp/include/ray/api/ray_config.h
Comment (SongGuyang) [reply cpp/include/ray/api/ray_config.h:22]: Why use two struct? Maybe one is enough?

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:13]:
```
// A list of directories or jar files that specify the search path for user
  // code. This will be used as `CLASSPATH` in Java, and `PYTHONPATH` in
  // Python.
```
The description in gcs.proto is for all languages.

**Comment by qicosmos** [comment]
> I didn't review library loading, function finding and execution part. I think @qicosmos already did a thorough review.

Yes, i did a review, but i think we need more test cases to cover the modifications.

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:64]: Good~

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:100]: ok

File: src/ray/raylet/worker_pool.cc
Comment (SongGuyang) [reply src/ray/raylet/worker_pool.cc:221]: Good~

File: cpp/include/ray/api/ray_config.h
Comment (kfstorm) [reply cpp/include/ray/api/ray_config.h:22]: Please refer to `ray.default.conf` for Java job config structures. And in Python, we use ray.init(..., job_config=ray.job_config.JobConfig(code_search_path=...)). So it's already two levels in both Java and Python.

File: cpp/src/ray/config_internal.cc
Comment (kfstorm) [reply cpp/src/ray/config_internal.cc:13]: You can remove the Java and Python parts.

File: cpp/include/ray/api/ray_config.h
Comment (SongGuyang) [reply cpp/include/ray/api/ray_config.h:22]: Python don't has RayConfig, so it should make a JobConfig to prevent a big number arguments of Ray.init function?

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:13]: Maybe you can show the description in your mind?

File: cpp/include/ray/api/ray_config.h
Comment (kfstorm) [reply cpp/include/ray/api/ray_config.h:22]: I don't think the two-level structure has anything to do with flattened parameters of `ray.init()` in Python. We designed the JobConfig class in Python and `ray.job.xxx` system property section in Java because we want to inform users that these options will have some impact on jobs, and they only impact jobs, not nodes or the cluster. That is, options designed to tune a job's behavior and have no impact on nodes or the cluster should be put in job config.

Comment (kfstorm) [reply cpp/include/ray/api/ray_config.h:22]: From another perspective, options in job config will not be shared by jobs. They are job-specific. Options not in job config are shared by all jobs. They are usually node or cluster options.

File: src/ray/protobuf/gcs.proto
Comment (kfstorm) [inline src/ray/protobuf/gcs.proto:288]:
```suggestion
  // A list of directories or files (jar files or dynamic libraries) that specify the search path
  // for user code. This will be used as `CLASSPATH` in Java, and `PYTHONPATH` in Python.
  // In C++, libraries in the path will be loaded by 'dlopen'.
```

File: cpp/src/ray/config_internal.cc
Comment (kfstorm) [reply cpp/src/ray/config_internal.cc:13]: A list of directories or files of dynamic libraries that specify the search path for user code. Libraries under these paths will be loaded by 'dlopen'.

File: cpp/include/ray/api/ray_config.h
Comment (SongGuyang) [reply cpp/include/ray/api/ray_config.h:22]: I think all the items in RayConfig should be job-specific.

**Review by kfstorm — APPROVED** [review]
LGTM

**Comment by kfstorm** [comment]
@AmeerHajAli @ericl Could you help review or approve? I cannot merge this PR because it requires code owner approval.

**Comment by AmeerHajAli** [comment]
@mwtian , is it possible for you to please review this?

File: cpp/src/ray/runtime/task/task_executor.cc
Comment (mwtian) [inline cpp/src/ray/runtime/task/task_executor.cc:30]: IIUC in `GetExecuteResult()` we look up `func_ptr` by `func_name`, then call `EntryFunction` which is this function? Are there other callsites that do not look up `func_ptr` by `func_name` before calling `TaskExecutionHandler()`? Should all callsites rely on `TaskExecutionHandler()` to look up `func_ptr` by `func_name` instead?

File: cpp/src/ray/runtime/task/task_executor.cc
Comment (mwtian) [inline cpp/src/ray/runtime/task/task_executor.cc:92]: I think we should avoid casting `RemoteFunction *` pointer returned from `GetExecutableFunctions()` to `void*`, then casting back to `RemoteFunction *`. This is error prone later on.

I think either we should rely on `EntryFuntion` to look up function by name, or use a wrapper like
```
class RemoteTask {
 public:
  explicit RemoteTask(RemoteFunction * func);
  RemoteTask(RemoteMemberFunction *func, msgpack::sbuffer *actor);
  void Run(const std::vector<msgpack::sbuffer> &args_buffer);
 private:
  ...
};
````
to wrap `actor_ptr`, `funcs` etc here?

**Review by mwtian — COMMENTED** [review]
IIUC the reason for this change is to use the same flag as Java and Python programs?

File: cpp/BUILD.bazel
Comment (SongGuyang) [reply cpp/BUILD.bazel:127]: I found sometimes the `cmd` in this `genrule` can not be executed. So I added this tag. And I found a lot of `genrule` in our repo have added this tag. What do you think about this?

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:42]: `code_search_path ` has been used everywhere in our repo. And there are the same named config in Python and Java API. Maybe we can modify this name in an independent PR?

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:52]: Good for me! Thanks.

File: cpp/src/ray/runtime/task/task_executor.cc
Comment (SongGuyang) [reply cpp/src/ray/runtime/task/task_executor.cc]: Good suggestion! I have refactored this part which use name to look up function. But I need to add a global map which help to find entry function from multiple libraries. The map is from function names to entry functions.

**Comment by SongGuyang** [comment]
@mwtian Thank you for careful review. I have addressed your comments. Please help to see this PR agent. Thank you.

File: cpp/include/ray/api/ray_config.h
Comment (SongGuyang) [reply cpp/include/ray/api/ray_config.h:22]: We will make another PR to do this. Issue https://github.com/ray-project/ray/issues/16994.

**Comment by SongGuyang** [comment]
@mwtian @AmeerHajAli No update? We need to merge this PR as soon as possible.

**Comment by AmeerHajAli** [comment]
Happy to approve once @mwtian approves

File: cpp/src/ray/config_internal.cc
Comment (raulchen) [reply cpp/src/ray/config_internal.cc:43]: @mwtian The single form "path" is already used by many software. E.g., "PATH" env var in shell, "classpath" in the Java compiler, and "include path" in C++ compilers. In their contexts, the term path can point to multiple directories as well, but the term is still in single form. So we decided to use the single form as well.

**Review by raulchen — APPROVED** [review]

File: cpp/src/ray/config_internal.cc
Comment (mwtian) [reply cpp/src/ray/config_internal.cc:43]: Keeping it consistent in our repo sounds good.

File: cpp/src/ray/test/cluster/counter.h
Comment (mwtian) [inline cpp/src/ray/test/cluster/counter.h:6]: Is this `using namespace` needed?
Also it might be better to wrap this in a namespace. Same for `plus.h`

**Comment by SongGuyang** [comment]
@mwtian Do you have any other update？We have milestone of C++ API and block in this PR. Can you update your comments or approve? Thanks.

File: cpp/src/ray/util/function_helper.cc
Comment (mwtian) [inline cpp/src/ray/util/function_helper.cc:109]: Why not accept `const std::vector<std::string> &paths` and avoid copying in this function?

File: cpp/src/ray/config_internal.h
Comment (mwtian) [inline cpp/src/ray/config_internal.h:30]: For the `Instance()` method below, can we return `const ConfigInternal &`?

**Review by mwtian — APPROVED** [review]
Structure wise this looks fine. High level design of C++ worker seems unchanged in this PR.

File: cpp/include/ray/api/function_manager.h
Comment (mwtian) [inline cpp/include/ray/api/function_manager.h:231]: It is better to never destruct instance, e.g. `static auto* instance = new FunctionManager()`. Otherwise `instance` destructor would run during program shutdown and maybe library unloading, which may cause hard to debug errors.

**Review by mwtian — APPROVED** [review]

**Review by AmeerHajAli — APPROVED** [review]
Thanks @mwtian for the detailed review, approving to unblock.

File: cpp/src/ray/util/function_helper.cc
Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.cc:109]: My fault. Thanks.

File: cpp/src/ray/config_internal.cc
Comment (SongGuyang) [reply cpp/src/ray/config_internal.cc:13]: Good for me.

File: cpp/src/ray/util/function_helper.cc
Comment (SongGuyang) [reply cpp/src/ray/util/function_helper.cc:55]: Sure, thanks.

File: cpp/include/ray/api/function_manager.h
Comment (SongGuyang) [reply cpp/include/ray/api/function_manager.h:231]: Good point! But do you mean make `FunctionHelper` never destructed? We save `boost::dll::shared_library` in `FunctionHelper`, not `FunctionManager `.

File: cpp/src/ray/config_internal.h
Comment (SongGuyang) [reply cpp/src/ray/config_internal.h:30]: I can't make it a const now because the config instance is mutable in current design. Maybe I can make it better in 
a future independent PR.

**Comment by SongGuyang** [comment]
@mwtian Thank you helpful comments! I have remove `tags = ["no-cache"]` which is not my hard request. And I have addressed your other comments. We will merge this PR today. If there are some other problems or issues, I can make another PR to fix.

File: cpp/src/ray/test/cluster/counter.h
Comment (SongGuyang) [reply cpp/src/ray/test/cluster/counter.h]: If users want to use some APIs in this library, the `using namespace` is needed. But now, I can delete this.

File: cpp/include/ray/api/function_manager.h
Comment (mwtian) [reply cpp/include/ray/api/function_manager.h:231]: Not destroying the container of `boost::dll::shared_library` is a great point. Anytime we need to have a static function local and it is not trivially destructible, it is recommended to store a pointer so it is never destructed. The argument is that otherwise if there is another thread reading the variable, the read can race with the static local variable's destruction.

File: cpp/src/ray/test/cluster/counter.h
Comment (mwtian) [reply cpp/src/ray/test/cluster/counter.h]: Ok. This is a test header so it does not matter that much. I think it is not recommended to have `using namespace` in a header especially not inside another namespace, unless the intention is to explicitly export those symbols with the header (e.g. https://google.github.io/styleguide/cppguide.html#Namespaces forbids `using namespace`).

**Pull request merged**
